llmfoundry/models/inference_api_wrapper/openai_causal_lm.py

```diff
@@ -0,0 +1,272 @@
+            cont_idxs = cont_idxs.tolist()
+            expected_cont_tokens = tokens[cont_idxs[0]:cont_idxs[-1] + 1]
+            output_logits = torch.nn.functional.one_hot(
+                torch.tensor(tokens[1:cont_idxs[0]]),
+                num_classes=self.tokenizer.vocab_size)
+
+            prompt = self.tokenizer.decode(tokens[:cont_idxs[0]])
+            next_logit_tensor = self.get_next_token_logit_tensor(
+                prompt, num_tokens=len(expected_cont_tokens))
+
+            if next_logit_tensor is not None:
+                output_logits = torch.cat([output_logits, next_logit_tensor])
+            padding = torch.nn.functional.one_hot(
+                torch.full((seqlen - output_logits.shape[0],),
+                           self.tokenizer.pad_token_id),
+                num_classes=self.tokenizer.vocab_size)
+            output_logits = torch.cat([output_logits, padding])
+            output_logits_batch.append(output_logits)
+
+        return torch.stack(output_logits_batch).to(batch['input_ids'].device)
+
+    def get_next_token_logit_tensor(self, prompt: str, num_tokens: int = 1):
+        tries = 0
+        chat_completion = None
+        while tries < MAX_RETRIES:
+            tries += 1
+            try:
+                chat_completion = openai.ChatCompletion.create(
+                    model=self.model_name,
+                    messages=[{
+                        'role': 'user',
+                        'content': prompt
+                    }],
+                    max_tokens=num_tokens,
+                    temperature=0.0)
+                break
+            except RateLimitError as e:
+                if 'You exceeded your current quota' in str(e._message):
+                    raise e
+                sleep(60)
+                continue
+            except Exception as e:
+                print(f'Caught exeception {str(e)}, continuing')
+                continue
+
+        if chat_completion is None:
+            raise ValueError("Couldn't generate model output")
+
+        assert isinstance(chat_completion, dict)
+        if len(chat_completion['choices']) > 0:
+            tensors = []
+            for t in self.tokenizer.encode(chat_completion['choices'][0]
+                                           ['message']['content'])['input_ids']:
+                tensors.append(
+                    self.tokenizer.construct_logit_tensor(
+                        {self.tokenizer.decode([t]): 0.0}))
+
+            if len(tensors) == 0:
+                return None
+            return torch.stack(tensors)
+        else:
+            # the model sometimes stops early even though we are still requesting tokens!
+            # not sure if there's a fix
+            return None
+
+
+class OpenAICausalLMEvalWrapper(InferenceAPIEvalWrapper):
+
+    def get_next_token_logit_tensor(self, prompt: str):
+        tries = 0
+        completion = None
+        while tries < MAX_RETRIES:
+            tries += 1
+            try:
```

Comment:
can you pull the try/except into a helper like `try_openai_request` since it is reused for both apis?